Avoid rewriting tiny matrices with cublas or triton

jaro-sevcik opened the pull request.

jaro-sevcik commented:
This is just a sketch of the approach for discussion. Problems with this version:
- Big one: *many* tests in gemm_rewrite_test fail because they mostly use small matrices. I do not see an easy workaround - they will have to be updated if go this way.
- we also might have to change cublas_pad_for_gemms to play along with this.
- missing tests for the actual change (that would be quite easy add).

nouiz commented on xla/service/gpu/matmul_utils.cc:
```diff
@@ -835,5 +835,43 @@ std::string TritonGemmConfig::ToString() const {
+                            dot_dims.lhs_contracting_dimensions()));
+  const int64_t rhs_non_contracting_size = absl::c_accumulate(
+      rhs_non_contracting_dims, 1, [&](int64_t size, int64_t dim) {
+        return size * lhs_shape.dimensions(dim);
```
lhs_shape -> rhs_shape?

nouiz replied:
Shouldn't this be:
absl::c_accumulate(rhs_non_contracting_dims, 1, std::multiplies<int64_t>());
instead?

jaro-sevcik replied:
> lhs_shape -> rhs_shape?

Good catch!

> Shouldn't this be:
> absl::c_accumulate(rhs_non_contracting_dims, 1, std::multiplies<int64_t>());
> instead?

I think it's correct as is, `rhs_non_contracting_dims` really contains dimension indexes, not sizes. The shape has the sizes.

nouiz commented:
@cheshire The PR is ready for review.

bchetioui commented on xla/debug_options_flags.cc:
```diff
@@ -222,6 +222,11 @@ DebugOptions DefaultDebugOptionsIgnoringFlags() {
 
+  // Minimum combined size of matrices in matrix multiplication to
+  // be rewritten to cuBLAS or Triton kernel call.
+  const int64_t kDefaultMinGemmRewriteSize = 100;
```
Why is that the right threshold value? Do we know that it's always better to use Triton/cuBLAS above this?

Not disagreeing, but I'd like documentation on why that is a good default (I'd also like to measure this over microbenchmarks & have this be part of the commit message to justify the change as well).

jaro-sevcik replied:
It turns out the threshold is quite conservative, see below. This was on H100 - on V100 the results are even more pronounced (0.015s vs 0.006s  at 16x16x16 in favor of skipped-rewriting).

I have also tested bf16, but there the numbers seem to be similar (10x  regardless of rewriting, presumably because of some padding.

```
H100 (1M float32)
  
   8 (2x2x2     ): SkipRewrite: 0.00008   Rewritten: 0.00238           
  12 (2x2x4     ): SkipRewrite: 0.00012   Rewritten: 0.00230
  12 (4x2x2     ): SkipRewrite: 0.00012   Rewritten: 0.00235
  16 (2x4x2     ): SkipRewrite: 0.00019   Rewritten: 0.00235
  16 (4x2x4     ): SkipRewrite: 0.00028   Rewritten: 0.00229
  20 (2x2x8     ): SkipRewrite: 0.00024   Rewritten: 0.00230
  20 (8x2x2     ): SkipRewrite: 0.00022   Rewritten: 0.00232
  24 (2x4x4     ): SkipRewrite: 0.00023   Rewritten: 0.00225
  24 (4x4x2     ): SkipRewrite: 0.00024   Rewritten: 0.00231
  24 (4x2x8     ): SkipRewrite: 0.00035   Rewritten: 0.00230
  24 (8x2x4     ): SkipRewrite: 0.00030   Rewritten: 0.00236
  32 (2x8x2     ): SkipRewrite: 0.00028   Rewritten: 0.00226
  32 (4x4x4     ): SkipRewrite: 0.00025   Rewritten: 0.00232
  32 (8x2x8     ): SkipRewrite: 0.00044   Rewritten: 0.00231
  36 (2x2x16    ): SkipRewrite: 0.00034   Rewritten: 0.00230
  36 (16x2x2    ): SkipRewrite: 0.00034   Rewritten: 0.00189
  40 (2x4x8     ): SkipRewrite: 0.00035   Rewritten: 0.00232
  40 (4x2x16    ): SkipRewrite: 0.00046   Rewritten: 0.00232
  40 (8x4x2     ): SkipRewrite: 0.00028   Rewritten: 0.00234
  40 (16x2x4    ): SkipRewrite: 0.00047   Rewritten: 0.00234
  48 (2x8x4     ): SkipRewrite: 0.00032   Rewritten: 0.00225
  48 (4x8x2     ): SkipRewrite: 0.00030   Rewritten: 0.00236
  48 (4x4x8     ): SkipRewrite: 0.00039   Rewritten: 0.00236
  48 (8x4x4     ): SkipRewrite: 0.00035   Rewritten: 0.00232
  48 (8x2x16    ): SkipRewrite: 0.00075   Rewritten: 0.00233
  48 (16x2x8    ): SkipRewrite: 0.00075   Rewritten: 0.00235
  64 (2x16x2    ): SkipRewrite: 0.00032   Rewritten: 0.00291
  64 (4x8x4     ): SkipRewrite: 0.00040   Rewritten: 0.00231
  64 (8x4x8     ): SkipRewrite: 0.00051   Rewritten: 0.00231
  64 (16x2x16   ): SkipRewrite: 0.00138   Rewritten: 0.00240
  72 (2x4x16    ): SkipRewrite: 0.00045   Rewritten: 0.00236
  72 (16x4x2    ): SkipRewrite: 0.00043   Rewritten: 0.00231
  80 (2x8x8     ): SkipRewrite: 0.00041   Rewritten: 0.00230
  80 (4x4x16    ): SkipRewrite: 0.00056   Rewritten: 0.00228
  80 (8x8x2     ): SkipRewrite: 0.00040   Rewritten: 0.00228
  80 (16x4x4    ): SkipRewrite: 0.00056   Rewritten: 0.00230
  96 (2x16x4    ): SkipRewrite: 0.00042   Rewritten: 0.00295
  96 (4x16x2    ): SkipRewrite: 0.00046   Rewritten: 0.00287
  96 (4x8x8     ): SkipRewrite: 0.00052   Rewritten: 0.00234
  96 (8x8x4     ): SkipRewrite: 0.00050   Rewritten: 0.00229
  96 (8x4x16    ): SkipRewrite: 0.00084   Rewritten: 0.00234
  96 (16x4x8    ): SkipRewrite: 0.00082   Rewritten: 0.00237
============================================================
 128 (4x16x4    ): SkipRewrite: 0.00059   Rewritten: 0.00289
 128 (8x8x8     ): SkipRewrite: 0.00070   Rewritten: 0.00236
 128 (16x4x16   ): SkipRewrite: 0.00138   Rewritten: 0.00250
 144 (2x8x16    ): SkipRewrite: 0.00065   Rewritten: 0.00238
 144 (16x8x2    ): SkipRewrite: 0.00061   Rewritten: 0.00237
 160 (2x16x8    ): SkipRewrite: 0.00064   Rewritten: 0.00286
 160 (4x8x16    ): SkipRewrite: 0.00077   Rewritten: 0.00238
 160 (8x16x2    ): SkipRewrite: 0.00062   Rewritten: 0.00294
 160 (16x8x4    ): SkipRewrite: 0.00078   Rewritten: 0.00233
 192 (4x16x8    ): SkipRewrite: 0.00079   Rewritten: 0.00292
 192 (8x16x4    ): SkipRewrite: 0.00082   Rewritten: 0.00294
 192 (8x8x16    ): SkipRewrite: 0.00113   Rewritten: 0.00239
 192 (16x8x8    ): SkipRewrite: 0.00113   Rewritten: 0.00240
 256 (8x16x8    ): SkipRewrite: 0.00108   Rewritten: 0.00296
 256 (16x8x16   ): SkipRewrite: 0.00184   Rewritten: 0.00250
 288 (2x16x16   ): SkipRewrite: 0.00110   Rewritten: 0.00288
 288 (16x16x2   ): SkipRewrite: 0.00103   Rewritten: 0.00293
 320 (4x16x16   ): SkipRewrite: 0.00125   Rewritten: 0.00289
 320 (16x16x4   ): SkipRewrite: 0.00127   Rewritten: 0.00297
 384 (8x16x16   ): SkipRewrite: 0.00179   Rewritten: 0.00292
 384 (16x16x8   ): SkipRewrite: 0.00181   Rewritten: 0.00292
 512 (16x16x16  ): SkipRewrite: 0.00264   Rewritten: 0.00302
 ```

jaro-sevcik replied:
For completeness, here are the bf16 numbers:

```
H100 (1M bf16 matmuls)
   8 (2x2x2     ): SkipRewrite: 0.00085   Rewritten: 0.00125
  12 (2x2x4     ): SkipRewrite: 0.00091   Rewritten: 0.00134
  12 (4x2x2     ): SkipRewrite: 0.00085   Rewritten: 0.00129
  16 (2x4x2     ): SkipRewrite: 0.00090   Rewritten: 0.00131
  16 (4x2x4     ): SkipRewrite: 0.00089   Rewritten: 0.00134
  20 (2x2x8     ): SkipRewrite: 0.00088   Rewritten: 0.00133
  20 (8x2x2     ): SkipRewrite: 0.00090   Rewritten: 0.00133
  24 (2x4x4     ): SkipRewrite: 0.00093   Rewritten: 0.00135
  24 (4x4x2     ): SkipRewrite: 0.00086   Rewritten: 0.00128
  24 (4x2x8     ): SkipRewrite: 0.00090   Rewritten: 0.00136
  24 (8x2x4     ): SkipRewrite: 0.00096   Rewritten: 0.00139
  32 (2x8x2     ): SkipRewrite: 0.00083   Rewritten: 0.00128
  32 (4x4x4     ): SkipRewrite: 0.00091   Rewritten: 0.00137
  32 (8x2x8     ): SkipRewrite: 0.00080   Rewritten: 0.00125
  36 (2x2x16    ): SkipRewrite: 0.00129   Rewritten: 0.00144
  36 (16x2x2    ): SkipRewrite: 0.00137   Rewritten: 0.00158
  40 (2x4x8     ): SkipRewrite: 0.00090   Rewritten: 0.00137
  40 (4x2x16    ): SkipRewrite: 0.00134   Rewritten: 0.00154
  40 (8x4x2     ): SkipRewrite: 0.00090   Rewritten: 0.00134
  40 (16x2x4    ): SkipRewrite: 0.00141   Rewritten: 0.00160
  48 (2x8x4     ): SkipRewrite: 0.00091   Rewritten: 0.00139
  48 (4x8x2     ): SkipRewrite: 0.00082   Rewritten: 0.00126
  48 (4x4x8     ): SkipRewrite: 0.00094   Rewritten: 0.00140
  48 (8x4x4     ): SkipRewrite: 0.00097   Rewritten: 0.00144
  48 (8x2x16    ): SkipRewrite: 0.00117   Rewritten: 0.00135
  48 (16x2x8    ): SkipRewrite: 0.00117   Rewritten: 0.00135
  64 (2x16x2    ): SkipRewrite: 0.00127   Rewritten: 0.00156
  64 (4x8x4     ): SkipRewrite: 0.00091   Rewritten: 0.00138
  64 (8x4x8     ): SkipRewrite: 0.00086   Rewritten: 0.00131
  64 (16x2x16   ): SkipRewrite: 0.00198   Rewritten: 0.00160
  72 (2x4x16    ): SkipRewrite: 0.00125   Rewritten: 0.00142
  72 (16x4x2    ): SkipRewrite: 0.00137   Rewritten: 0.00161
  80 (2x8x8     ): SkipRewrite: 0.00065   Rewritten: 0.00114
  80 (4x4x16    ): SkipRewrite: 0.00134   Rewritten: 0.00152
  80 (8x8x2     ): SkipRewrite: 0.00069   Rewritten: 0.00117
  80 (16x4x4    ): SkipRewrite: 0.00142   Rewritten: 0.00159
  96 (2x16x4    ): SkipRewrite: 0.00128   Rewritten: 0.00155
  96 (4x16x2    ): SkipRewrite: 0.00129   Rewritten: 0.00155
  96 (4x8x8     ): SkipRewrite: 0.00069   Rewritten: 0.00116
  96 (8x8x4     ): SkipRewrite: 0.00072   Rewritten: 0.00117
  96 (8x4x16    ): SkipRewrite: 0.00119   Rewritten: 0.00136
  96 (16x4x8    ): SkipRewrite: 0.00123   Rewritten: 0.00137
 128 (4x16x4    ): SkipRewrite: 0.00131   Rewritten: 0.00159
 128 (8x8x8     ): SkipRewrite: 0.00042   Rewritten: 0.00092
 128 (16x4x16   ): SkipRewrite: 0.00203   Rewritten: 0.00171
 144 (2x8x16    ): SkipRewrite: 0.00100   Rewritten: 0.00116
 144 (16x8x2    ): SkipRewrite: 0.00111   Rewritten: 0.00129
 160 (2x16x8    ): SkipRewrite: 0.00099   Rewritten: 0.00129
 160 (4x8x16    ): SkipRewrite: 0.00109   Rewritten: 0.00126
 160 (8x16x2    ): SkipRewrite: 0.00105   Rewritten: 0.00132
 160 (16x8x4    ): SkipRewrite: 0.00115   Rewritten: 0.00132
 192 (4x16x8    ): SkipRewrite: 0.00102   Rewritten: 0.00130
 192 (8x16x4    ): SkipRewrite: 0.00107   Rewritten: 0.00133
 192 (8x8x16    ): SkipRewrite: 0.00074   Rewritten: 0.00092
 192 (16x8x8    ): SkipRewrite: 0.00077   Rewritten: 0.00093
 256 (8x16x8    ): SkipRewrite: 0.00063   Rewritten: 0.00095
 256 (16x8x16   ): SkipRewrite: 0.00140   Rewritten: 0.00099
 288 (2x16x16   ): SkipRewrite: 0.00149   Rewritten: 0.00138
 288 (16x16x2   ): SkipRewrite: 0.00153   Rewritten: 0.00151
 320 (4x16x16   ): SkipRewrite: 0.00159   Rewritten: 0.00148
 320 (16x16x4   ): SkipRewrite: 0.00158   Rewritten: 0.00154
 384 (8x16x16   ): SkipRewrite: 0.00108   Rewritten: 0.00097
 384 (16x16x8   ): SkipRewrite: 0.00104   Rewritten: 0.00097
 512 (16x16x16  ): SkipRewrite: 0.00181   Rewritten: 0.00128
```

bchetioui replied:
Thank you so much for taking the time to benchmark this! Should we include these results in the commit message for traceability?

I also wonder what we could put in the code to document this threshold. Would it be reasonable to add something to the comment that says something like "This threshold is a conservative estimate and has been measured to be always beneficial (up to generally several times faster) on V100 and H100 GPUs. See openxla/xla #9319 for details."?

bchetioui commented on xla/service/gpu/gemm_rewriter_triton_test.cc:
```diff
@@ -1097,6 +1098,64 @@ ENTRY e {
+})");
+}
+
+TEST_F(SmallMatrixGemmRewriterTritonTest, LargeMatrixRewrite) {
```
nit: maybe "[Perform]LargeMatrixRewrite" or "RewriteLargeMatrix" would be better, to make it clear we are checking that a large dot is being rewritten (otherwise "Rewrite" reads like a noun instead of a verb).

bchetioui commented on xla/service/gpu/gemm_rewriter_triton_test.cc:
```diff
@@ -1097,6 +1098,64 @@ ENTRY e {
+  }
+};
+
+TEST_F(SmallMatrixGemmRewriterTritonTest, SkipSmallMatrixRewrite) {
```
nit: "Matrix" could be "Dot" (or something to that effect) here and below? What's (may be getting) rewritten is not the Matrix, but the Dot operation.

bchetioui commented on xla/debug_options_flags.cc:
```diff
@@ -1463,6 +1468,12 @@ void MakeDebugOptionsFlags(std::vector<tsl::Flag>* flag_list,
+      int64_setter_for(&DebugOptions::set_xla_gpu_gemm_rewrite_size_threshold),
+      debug_options->xla_gpu_gemm_rewrite_size_threshold(),
+      "Threshold to rewrite matmul to cuBLAS or Triton "
+      "(total number of elements)."));
```
This documentation does not seem to correspond to the implementation: it seems like the implementation is ignoring batch dimensions---thus not the total number of elements. Let's document exactly how this threshold relates to the dot operands.

bchetioui commented on xla/service/gpu/matmul_utils.h:
```diff
@@ -70,6 +70,10 @@ absl::StatusOr<Shape> GetBatchRowColumnShape(
 absl::StatusOr<bool> CanFoldTransposeOperandIntoDot(const HloInstruction& dot,
                                                     int64_t operand_idx);
 
+// Returns true if the matrices are too small to benefit from cublas/triton.
```
This docstring is misleading; this will return true if the size of each (unbatched) matmul is below the threshold, independently of whether cuBLAS and Triton would be good codegen solutions.

Maybe "Returns true if the combined size of the (unbatched) operands is smaller than the given threshold."?

bchetioui commented on xla/service/gpu/matmul_utils_test.cc:
```diff
@@ -225,6 +225,80 @@ TEST(GetMatrixLayoutTest, BatchInMostMinorPhysicalDimension) {
+ENTRY DotFunc {
+  x = f32[100,30,3] parameter(0)
+  y = f32[100,3,3] parameter(1)
+  ROOT dot_a = f32[100,30,3] dot(x, y), lhs_contracting_dims={2}, rhs_contracting_dims={1}, lhs_batch_dims={0}, rhs_batch_dims={0}
```
nit: all dot roots can just have name `dot`

bchetioui commented on xla/service/gpu/matmul_utils_test.cc:
```diff
@@ -225,6 +225,80 @@ TEST(GetMatrixLayoutTest, BatchInMostMinorPhysicalDimension) {
   EXPECT_FALSE(MatrixLayout::For(shape).ok());
 }
 
+using GetMatrixSizeRewriteThresholdTest = HloTestBase;
```
Are we missing a test that batch dimensions are ignored here?

bchetioui replied:
Nevermind there is one below---I didn't see :)

bchetioui commented on xla/service/gpu/tests/gemm_rewrite_test.cc:
```diff
@@ -7127,6 +7135,59 @@ ENTRY AddDotsFunc {
+}
+)";
+
+  EXPECT_TRUE(RunAndCompare(hlo_text, ErrorSpec{1e-5, 1e-5}));
```
nit: not sure if `RunAndCompare`s are necessary here (these should already be covered by other tests, either in the elemental rewriter or above in the same file for cuBLAS)

bchetioui reviewed (commented):
Thanks for the changes! Sorry, I realize that I posted my comments one by one instead of sending a full review. I don't use GitHub enough :-))

jaro-sevcik reviewed (commented):
Thank you for the review! I think I have addressed all the code suggestions, but I still have not finished the benchmarks that justify this. I am hoping to be done with them tomorrow.

bchetioui reviewed (commented):
Thank you @jaro-sevcik for the changes and for thoughtfully addressing the comments! I'll be waiting for your update regarding setting and documenting the threshold, but after that I think we can merge this!

bchetioui commented:
Friendly ping regarding [my last comment](https://github.com/openxla/xla/pull/9319#discussion_r1492465193)---in case you have missed it @jaro-sevcik :)

jaro-sevcik commented:
I fixed the tests, enabled the cublas/triton bypass only for floats (since some integer cases crash) and added the comment you suggested. Could you (@bchetioui) take another look, please?

bchetioui commented on xla/service/gpu/matmul_utils.cc:
```diff
@@ -787,5 +787,53 @@ std::string TritonGemmConfig::ToString() const {
+         threshold;
+}
+
+bool IsDotSupportedByBackend(const HloInstruction& dot) {
```
Thanks! Maybe `IsDotSupportedByClassicalEmitters`? `Backend` is not extremely clear.

I'd also suggest making a GitHub issue with your findings and linking it here so that the lack of support for some int matmuls does not get lost :)

jaro-sevcik replied:
Done.

bchetioui commented on xla/service/gpu/matmul_utils_test.cc:
```diff
@@ -225,6 +225,80 @@ TEST(GetMatrixLayoutTest, BatchInMostMinorPhysicalDimension) {
   EXPECT_FALSE(MatrixLayout::For(shape).ok());
 }
 
+using GetMatrixSizeRewriteThresholdTest = HloTestBase;
```
Should this file now also have an additional test to make sure that the matrix multiplication is also rewritten if the data type is supported?

jaro-sevcik replied:
I have added some tests, but I am not quite sure of their value.

bchetioui replied:
I think I would have checked that the matrix multiplication does not end up being codegen'd by old emitters if we expect it not to be covered correctly. I can make this change later to avoid the back and forth for you, since this seems good enough now :)

bchetioui reviewed (commented):
Thank you @jaro-sevcik. Your approach for int matrices makes sense to me :)

jaro-sevcik reviewed (commented):
Addressed the comments. Could you take a look?

bchetioui reviewed (approved):
Thank you for thoughtfully addressing all my comments @jaro-sevcik!

bchetioui commented:
FYI @jaro-sevcik, this is blocked on some tests that need updating due to the newly added paths (and some numerical tolerances that need to be bumped).

I plan to fix them myself and land this whenever I get some bandwidth (probably today, otherwise early next week).

jaro-sevcik commented:
Thank you for doing all the extra work, we are not in a rush.

bchetioui commented:
Just to update you, I spent a couple hours fixing this for landing today. Unfortunately, it's breaking a test for which I haven't yet figured what the best way to rewrite it is---so it may take a teeny bit more time.

jaro-sevcik commented:
Than you for the update. Please, let me know if I can help in any way.

tdanyluk commented:
Note (mainly to @bchetioui and myself): Once explicit algorithm support lands, we will have to change this so that we only avoid the rewrite if the "classical" emitter supports the given algorithm.

bchetioui commented:
Thanks for your patience @jaro-sevcik. This should be landing now.
